lambdas/access_counts/index.py

```diff
@@ -0,0 +1,161 @@
+"""
+Lambda function that runs Athena queries over CloudTrail logs and .quilt/named_packages/
+and creates summaries of object and package access events.
+"""
+
+import csv
+from collections import defaultdict
+import json
+import os
+import time
+from tempfile import TemporaryFile
+
+import boto3
+
+# A saved query that reduces CloudTrail logs event counts per object per day.
+OBJECT_ACCESS_COUNTS_QUERY_ID = os.environ['OBJECT_ACCESS_COUNTS_QUERY_ID']
+# Named packages
+NAMED_PACKAGES_QUERY_ID = os.environ['NAMED_PACKAGES_QUERY_ID']
+# Bucket where query results will be stored.
+QUERY_RESULT_BUCKET = os.environ['QUERY_RESULT_BUCKET']
+# A temporary directory where Athena query results will be written.
+QUERY_TEMP_DIR = os.environ['QUERY_TEMP_DIR']
+# Directory where the summary files will be stored.
+ACCESS_COUNTS_OUTPUT_DIR = os.environ['ACCESS_COUNTS_OUTPUT_DIR']
+
+
+MAX_RESULTS = 1000
+
+
+athena = boto3.client('athena')
+s3 = boto3.client('s3')
+
+
+def run_query(named_query_id):
+    output = 's3://%s/%s/' % (QUERY_RESULT_BUCKET, QUERY_TEMP_DIR)
+
+    query = athena.get_named_query(NamedQueryId=named_query_id)['NamedQuery']
+    response = athena.start_query_execution(
+        QueryString=query['QueryString'],
+        QueryExecutionContext=dict(Database=query['Database']),
+        ResultConfiguration=dict(OutputLocation=output)
+    )
+    print("Started query:", response)
+
+    execution_id = response['QueryExecutionId']
+
+    return execution_id
+
+
+def wait_for_query(execution_id):
+    while True:
+        response = athena.get_query_execution(QueryExecutionId=execution_id)
+        print("Query status:", response)
+        state = response['QueryExecution']['Status']['State']
+
+        if state == 'RUNNING':
+            pass
+        elif state == 'SUCCEEDED':
+            break
+        elif state == 'FAILED':
+            raise Exception("Query failed! QueryExecutionId=%r" % execution_id)
+        elif state == 'CANCELLED':
+            raise Exception("Query cancelled! QueryExecutionId=%r" % execution_id)
+        else:
+            assert False, "Unexpected state: %s" % state
+
+        time.sleep(5)
+
+
+def get_query_results(execution_id):
+    args = dict(
+        QueryExecutionId=execution_id,
+        MaxResults=MAX_RESULTS
+    )
+
+    header = True
+
+    while True:
+        response = athena.get_query_results(**args)
+        for row in response['ResultSet']['Rows']:
+            if header:
+                # Athena returns the header before the actual results. Sigh.
+                header = False
+            else:
+                yield [value['VarCharValue'] for value in row['Data']]
```

Comment:
it would be super helpful to document (in this file) the expected schemas of all underlying files. for instance, i was looking at https://github.com/quiltdata/t4/pull/546/files#diff-76a245204f5971faf14b2f56e0b84da4R130 and i have no clear sense of whether that line actually matches the data we're returning.